<a href="https://colab.research.google.com/github/jigsawanku/Hotel-booking/blob/main/Hotel_booking_analysis_Ankit_Garg.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
# Importing Liberaries which will be used further to explore the data
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [3]:
df1 = pd.read_csv("/content/drive/MyDrive/AlmaBetter/Hotel_Booking_Analysis - Ankit Garg/Hotel Bookings.csv")

In [4]:
df1.head()

hotel  is_canceled  ...  reservation_status  reservation_status_date
0  Resort Hotel            0  ...           Check-Out               2015-07-01
1  Resort Hotel            0  ...           Check-Out               2015-07-01
2  Resort Hotel            0  ...           Check-Out               2015-07-02
3  Resort Hotel            0  ...           Check-Out               2015-07-02
4  Resort Hotel            0  ...           Check-Out               2015-07-03

[5 rows x 32 columns]

In [5]:
df1.tail()

hotel  is_canceled  ...  reservation_status  reservation_status_date
119385  City Hotel            0  ...           Check-Out               2017-09-06
119386  City Hotel            0  ...           Check-Out               2017-09-07
119387  City Hotel            0  ...           Check-Out               2017-09-07
119388  City Hotel            0  ...           Check-Out               2017-09-07
119389  City Hotel            0  ...           Check-Out               2017-09-07

[5 rows x 32 columns]

In [7]:
df1.shape

(119390, 32)

In [9]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [10]:
# checking the total number of missing values available in each variable
df1.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

In [11]:
# Converting the total number of missing values available in each variable into percentage
df1_null_perctange = df1.isnull().sum()/df1.shape[0]*100
df1_null_perctange

hotel                              0.000000
is_canceled                        0.000000
lead_time                          0.000000
arrival_date_year                  0.000000
arrival_date_month                 0.000000
arrival_date_week_number           0.000000
arrival_date_day_of_month          0.000000
stays_in_weekend_nights            0.000000
stays_in_week_nights               0.000000
adults                             0.000000
children                           0.003350
babies                             0.000000
meal                               0.000000
country                            0.408744
market_segment                     0.000000
distribution_channel               0.000000
is_repeated_guest                  0.000000
previous_cancellations             0.000000
previous_bookings_not_canceled     0.000000
reserved_room_type                 0.000000
assigned_room_type                 0.000000
booking_changes                    0.000000
deposit_type                    

In [12]:
drop_column = df1_null_perctange[df1_null_perctange>20].keys()
drop_column

Index(['company'], dtype='object')

In [13]:
# Dropping the columns which has more than 20% of missing values
df1_without_null = df1.drop(columns=drop_column)
df1_without_null.shape

(119390, 31)

In [16]:
# imputing missing values of variable "average" with '0' 
df1_without_null['agent']=df1_without_null['agent'].fillna(np.mean(df1_without_null['agent']))

In [17]:
df1_without_null.isnull().sum()

hotel                               0
is_canceled                         0
lead_time                           0
arrival_date_year                   0
arrival_date_month                  0
arrival_date_week_number            0
arrival_date_day_of_month           0
stays_in_weekend_nights             0
stays_in_week_nights                0
adults                              0
children                            4
babies                              0
meal                                0
country                           488
market_segment                      0
distribution_channel                0
is_repeated_guest                   0
previous_cancellations              0
previous_bookings_not_canceled      0
reserved_room_type                  0
assigned_room_type                  0
booking_changes                     0
deposit_type                        0
agent                               0
days_in_waiting_list                0
customer_type                       0
adr         

In [18]:
df1_without_null=df1_without_null.dropna()

#To insure we don't have any null values
df1_without_null.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces 

In [19]:
df_duplicate_rows = df1_without_null[df1_without_null.duplicated()].shape

print(f"the total no. of duplicate rows :" , df_duplicate_rows)

the total no. of duplicate rows : (31965, 31)


In [22]:
df2=df1_without_null.drop_duplicates()
df2.shape

(86933, 31)

In [30]:
df2.dtypes

hotel                              object
is_canceled                         int64
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                 object
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                               object
country                            object
market_segment                     object
distribution_channel               object
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
reserved_room_type                 object
assigned_room_type                 object
booking_changes                     int64
deposit_type                       object
agent                             

In [31]:
df2.head(2)

hotel  is_canceled  ...  reservation_status  reservation_status_date
0  Resort Hotel            0  ...           Check-Out               2015-07-01
1  Resort Hotel            0  ...           Check-Out               2015-07-01

[2 rows x 31 columns]

In [33]:
# Getting the unique values of all the categoricals variables which  has less than 50 unique values
unique_list_long = []
unique_list_short = []

for col in df2.columns:
  
  uni_list = len(df2[col].unique())
  if uni_list>50:
    unique_list_long.append('unique_elements of '+col+' is '+ str(df2[col].unique()))
  else:
    unique_list_short.append('unique_elements of '+col+' is '+ str(df2[col].unique()))
print(unique_list_short)

["unique_elements of hotel is ['Resort Hotel' 'City Hotel']", 'unique_elements of is_canceled is [0 1]', 'unique_elements of arrival_date_year is [2015 2016 2017]', "unique_elements of arrival_date_month is ['July' 'August' 'September' 'October' 'November' 'December' 'January'\n 'February' 'March' 'April' 'May' 'June']", 'unique_elements of arrival_date_day_of_month is [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24\n 25 26 27 28 29 30 31]', 'unique_elements of stays_in_weekend_nights is [ 0  1  2  4  3  6 13  8  5  7 12  9 16 10 14]', 'unique_elements of stays_in_week_nights is [ 0  1  2  3  4  5 11  8 10  6  7 15  9 12 33 20 14 16 21 13 30 19 24 40\n 22 25 17 32 26 18 34 35 41]', 'unique_elements of adults is [ 2  1  3  4 40 26 50 27 55  0 20  6  5 10]', 'unique_elements of children is [ 0.  1.  2. 10.  3.]', 'unique_elements of babies is [ 0  1  2 10  9]', "unique_elements of meal is ['BB' 'FB' 'HB' 'SC' 'Undefined']", "unique_elements of market_segment is [

In [34]:
df2.describe()

is_canceled  ...  total_of_special_requests
count  86933.000000  ...               86933.000000
mean       0.275925  ...                   0.699481
std        0.446982  ...                   0.831845
min        0.000000  ...                   0.000000
25%        0.000000  ...                   0.000000
50%        0.000000  ...                   0.000000
75%        1.000000  ...                   1.000000
max        1.000000  ...                   5.000000

[8 rows x 19 columns]

*   27.59 % of the people have cancelled their booking as per the dataset.
*   Avg. lead time is 80 days.
*   Only 4% of the guests are repeated.
*   Each booking has on an average 1.8 adults and 0.13 children.